src/rtc_stream.cpp

```diff
@@ -0,0 +1,418 @@
+    }
+
+	auto addr = m_peer_connection->localAddress();
+	if (!addr)
+	{
+		ec = boost::asio::error::operation_not_supported;
+		return {};
+	}
+
+	// The format is always "address:port" without brackets around the address,
+	// therefore splitting on the last ':' works to separate address and port.
+	std::size_t pos = addr->find_last_of(':');
+	if (pos == std::string::npos)
+	{
+		ec = errors::parse_failed;
+		return {};
+	}
+
+	return endpoint_type(ip::make_address(addr->substr(0, pos), ec)
+			, std::uint16_t(std::stoul(addr->substr(pos+1))));
+}
+
+void rtc_stream_impl::cancel_handlers(error_code const& ec)
+{
+	TORRENT_ASSERT(ec);
+
+	auto read_handler = std::exchange(m_read_handler, nullptr);
+	auto write_handler = std::exchange(m_write_handler, nullptr);
+
+	m_read_handler = nullptr;
+	m_read_buffer.clear();
+	m_read_buffer_size = 0;
+
+	m_write_handler = nullptr;
+	m_write_buffer.clear();
+	m_write_buffer_size = 0;
+
+	if (read_handler) read_handler(ec, 0);
+	if (write_handler) write_handler(ec, 0);
+}
+
+bool rtc_stream_impl::ensure_open()
+{
+	if (is_open()) return true;
+
+    cancel_handlers(boost::asio::error::not_connected);
+    return false;
+}
+
+
+void rtc_stream_impl::issue_read()
+{
+	TORRENT_ASSERT(m_read_handler);
+	TORRENT_ASSERT(m_read_buffer_size > 0);
+
+	if (!ensure_open()) return;
+
+	error_code ec;
+	std::size_t bytes_read = read_some(ec);
+	if (ec || bytes_read > 0) // error or immediate read
+	{
+		clear_read_buffers();
+		post(m_io_context, std::bind(std::exchange(m_read_handler, nullptr), ec, bytes_read));
+	}
+}
+
+void rtc_stream_impl::issue_write()
+{
+	TORRENT_ASSERT(m_write_handler);
+	TORRENT_ASSERT(m_write_buffer_size > 0);
+
+	if (!ensure_open()) return;
+
+	std::size_t const max_message_size = m_data_channel->maxMessageSize();
+	std::size_t bytes_written = 0;
+	bool is_buffered = false;
+	while (!m_write_buffer.empty())
+	{
+		bytes_written += write_data(max_message_size, is_buffered);
+	}
+
+	TORRENT_ASSERT(bytes_written == m_write_buffer_size);
+
+	if (!is_buffered)
+	{
+		m_write_buffer.clear();
+		m_write_buffer_size = 0;
+		post(m_io_context, std::bind(std::exchange(m_write_handler, nullptr), error_code{}, bytes_written));
+	}
+}
+
+std::size_t rtc_stream_impl::read_some(error_code& ec)
+{
+	if (!is_open())
+	{
+		ec = boost::asio::error::not_connected;
+		return 0;
+	}
+
+	std::size_t bytes_read = 0;
+
+	if (!m_incoming.empty())
+	{
+		std::size_t const copied = read_data(m_incoming.data(), m_incoming.size());
+		bytes_read += copied;
+		if (copied < m_incoming.size())
+		{
+			m_incoming.erase(m_incoming.begin(), m_incoming.begin() + std::ptrdiff_t(copied));
+			return bytes_read;
+		}
+
+		m_incoming.clear();
+	}
+
+	while (!m_read_buffer.empty() && m_incoming.empty() && !ec)
+	{
+		auto message = m_data_channel->receive();
+		if (!message) break;
+
+		std::visit(rtc::overloaded
+		{
+			[&](rtc::binary const& bin)
+			{
+				char const *data = reinterpret_cast<char const*>(bin.data());
```

Comment:
just as a suggestion, if `rtc::binary` would be something similar to `boost::asio::const_buffer`, and represent a buffer as a `void const*`, it would be sufficient with a `static_cast<>` here, instead of a `reinterpret_cast<>`.

Reply:
Yes, libdatachannel use the type `std::byte` everywhere for binary data so for consistency I'd like to keep `byte` there too.